src/go/debug-collector/dumper/dumper.go

```diff
@@ -0,0 +1,222 @@
+package dumper
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// Dumper struct is for dumping cluster
+type Dumper struct {
+	cmd       string
+	resources []string
+	namespace string
+	location  string
+	Errors    map[string]string
+	mode      int64
+}
+
+// New return new Dumper object
+func New(location, namespace, resource string) Dumper {
+	directory := "cluster-dump"
+	if len(location) > 0 {
+		directory = location + "/cluster-dump"
+	}
+	resources := []string{
+		"pods",
+		"replicasets",
+		"deployments",
+		"statefulsets",
+		"replicationcontrollers",
+		"events",
+		"configmaps",
+		"secrets",
+		"cronjobs",
+		"jobs",
+		"podsecuritypolicies",
+		"poddisruptionbudgets",
+		"perconaxtradbbackups",
+		"perconaxtradbclusterbackups",
+		"perconaxtradbclusterrestores",
+		"perconaxtradbclusters",
+		"clusterrolebindings",
+		"clusterroles",
+		"rolebindings",
+		"roles",
+		"storageclasses",
+		"persistentvolumeclaims",
+		"persistentvolumes",
+	}
+	if len(resource) > 0 {
+		resources = append(resources, resource)
+	}
+	return Dumper{
+		cmd:       "kubectl",
+		resources: resources,
+		location:  directory,
+		Errors:    make(map[string]string),
+		mode:      int64(0777),
+		namespace: namespace,
+	}
+}
+
+type k8sPods struct {
+	Items []corev1.Pod `json:"items"`
+}
+
+type namespaces struct {
+	Items []corev1.Namespace `json:"items"`
+}
+
+// DumpCluster create dump of a cluster in Dumper.location
+func (d *Dumper) DumpCluster() error {
+	tarFile, err := os.Create(d.location + ".tar.gz")
+	if err != nil {
+		return errors.Wrap(err, "create tar file")
+	}
+	defer tarFile.Close()
+	zr := gzip.NewWriter(tarFile)
+	tw := tar.NewWriter(zr)
+	defer zr.Close()
+	defer tw.Close()
+
+	var nss namespaces
+
+	if len(d.namespace) > 0 {
+		ns := corev1.Namespace{}
+		ns.Name = d.namespace
+		nss.Items = append(nss.Items, ns)
+	} else {
+		output, err := d.runCmd("get", "namespaces", "-o", "json")
+		if err != nil {
+			return errors.Wrap(err, "get namespaces")
+		}
+
+		err = json.Unmarshal(output, &nss)
+		if err != nil {
+			return errors.Wrap(err, "unmarshal namespaces")
+		}
+	}
+
+	for _, ns := range nss.Items {
+		output, err := d.runCmd("get", "pods", "-o", "json", "--namespace", ns.Name)
+		if err != nil {
+			continue // runCmd already stored this error in Dumper.Errors
+		}
+		var pods k8sPods
+		err = json.Unmarshal(output, &pods)
+		if err != nil {
+			d.saveCommandError(err.Error(), "unmarshal pods from namespace", ns.Name)
+			log.Println(errors.Wrap(err, "unmarshal pods"))
+		}
+
+		for _, pod := range pods.Items {
+			location := d.location + "/" + ns.Name + "/" + pod.Name + "/logs.txt"
+			output, err = d.runCmd("logs", pod.Name, "--namespace", ns.Name, "--all-containers")
+			if err != nil {
+				err = createArchive(location, d.mode, []byte(err.Error()), tw)
```

Comment:
You just going to write only an error returned by `exec.Run` into the logs.txt. And into `errors.txt` you are going to write stderr & stdout + sometimes the error. But the content that lands in both `logs.txt` and `errors.txt` should be the same

Reply:
https://github.com/percona/percona-toolkit/pull/452/commits/118110a671ae050d9ba09342b631a0fecf9a964e